index.ts

```diff
@@ -0,0 +1,529 @@
+        this.options = options;
+        this.initialized = false;
+
+        this.commands = {
+            create_domain: {
+                lifecycleEvents: [
+                    "create",
+                    "initialize",
+                ],
+                usage: "Creates a domain using the domain name defined in the serverless file",
+            },
+            delete_domain: {
+                lifecycleEvents: [
+                    "delete",
+                    "initialize",
+                ],
+                usage: "Deletes a domain using the domain name defined in the serverless file",
+            },
+        };
+        this.hooks = {
+            "after:deploy:deploy": this.setupBasePathMapping.bind(this),
+            "after:info:info": this.domainSummary.bind(this),
+            "before:remove:remove": this.removeBasePathMapping.bind(this),
+            "create_domain:create": this.createDomain.bind(this),
+            "delete_domain:delete": this.deleteDomain.bind(this),
+        };
+    }
+
+    /**
+     * Lifecycle function to create a domain
+     * Wraps creating a domain and resource record set
+     */
+    public async createDomain() {
+        this.initializeVariables();
+        if (!this.enabled) {
+            this.reportDisabled();
+            return;
+        }
+        const certArn = await this.getCertArn();
+        const domainInfo = await this.createCustomDomain(certArn);
+        await this.changeResourceRecordSet("UPSERT", domainInfo);
+        this.serverless.cli.log(`Custom domain ${this.givenDomainName} was created/updated.
+      New domains may take up to 40 minutes to be initialized.`);
+    }
+
+    /**
+     * Lifecycle function to delete a domain
+     * Wraps deleting a domain and resource record set
+     */
+    public async deleteDomain() {
+        this.initializeVariables();
+        if (!this.enabled) {
+            this.reportDisabled();
+            return;
+        }
+        const domainInfo = await this.getDomainInfo();
+        await this.deleteCustomDomain();
+        await this.changeResourceRecordSet("DELETE", domainInfo);
+        this.serverless.cli.log(`Custom domain ${this.givenDomainName} was deleted.`);
+    }
+
+    /**
+     * Lifecycle function to create basepath mapping
+     * Wraps creation of basepath mapping and adds domain name info as output to cloudformation stack
+     */
+    public async setupBasePathMapping() {
+        this.initializeVariables();
+        if (!this.enabled) {
+            this.reportDisabled();
+            return;
+        }
+        await this.createBasePathMapping();
+        const domainInfo = await this.getDomainInfo();
+        this.addOutputs(domainInfo);
+        await this.printDomainSummary(domainInfo);
+    }
+
+    /**
+     * Lifecycle function to delete basepath mapping
+     * Wraps deletion of basepath mapping
+     */
+    public async removeBasePathMapping() {
+        this.initializeVariables();
```

Comment:
Try having a decorator function to make these calls that takes in the function below?